client/network/src/config.rs

```diff
@@ -605,6 +606,9 @@ pub struct NetworkConfiguration {
 	/// the presence of potentially adversarial nodes.
 	pub kademlia_disjoint_query_paths: bool,
 
+	/// None will end up using default Kademlia value.
+	pub kademlia_replication_factor: Option<NonZeroUsize>,
```

Comment:
Why does this need to be an `Option` if the default value is provided in `network_params.rs`?

Reply:
Because Kademlia libp2p `NetworkBehaviour` doesn't need to configured with any replication factor to work.

See [this](https://github.com/libp2p/rust-libp2p/blob/b8ceeccdc6a4d38ebe77d74cab3c66be072f16db/protocols/kad/src/query.rs#L253). Having `Option` here relieves use the burden of investigating and figuring out what value should be set for `kademlia_replication_factor`.

Makes sense?

Reply:
Then we'd also have to default value [here](https://github.com/paritytech/substrate/pull/14391/files#diff-390c55bb9dcff07a1ee409bae9f4cdea9a6f21cf438090e6f25009bc63858d86R663).

There's also question of what if libp2p's `K_VALUE` changes and we've defaulted to 20 as default value for corresponding CLI flag. Do we want to be in sync with what libp2p believes to be best default value for Kademlia or not?

1. If you think we should just use `20` as default value for this and there's no need to be in sync if libp2p changes it then I'll remove `Option`.

2. Otherwise let's just keep `Option` in CLI flag:

```rust
#[arg(long, default_value = "20")]
pub kademlia_replication_factor: NonZeroUsize,
```

to this

```rust
#[arg(long)]
pub kademlia_replication_factor: Option<NonZeroUsize>,
```

and default to `K_VALUE` if none was provided. Note that in this case we can't really display what is default value in `CLI --help`, only refer to libp2p.

I'm ok with doing both 1. and 2. I'll let you decide.

Reply:
The replication factor is mentioned in the spec so there's definitely merit in having a default value and removing the `Option` route. But that would mean that if Protocol Labs/some other entity would find a better replication factor we'd have to be aware of this research and also update our default value accordingly.

I'd lean towards removing the default value and having it as `Option<NonZeroUsize>` as a CLI parameter and in the absence of user-defined replication factor, we use whatever the Kademlia provider deems best. Any thoughts @bkchr?

Reply:
It is always nice to see the default value in CI, but I'm also don't have a hard requirement on this. Not sure the default value by Protocol labs would change any way. That said, do whatever you think is the best @altonen :D

Reply:
Alright, let's keep the default value and remove the `Option`.